utils/contract.ts

```diff
@@ -0,0 +1,90 @@
+import axios from "axios";
+import { ETHERSCAN_API_KEY, execute } from ".";
+import { type SubmissionConfig } from "../types";
+import fs from "fs";
+
+/**
+ * 1. Fetch the contract source code from Etherscan
+ * 2. Save the contract source code into the appropriate challenge directory
+ */
+export const downloadContract = async (
+  config: SubmissionConfig
+): Promise<void> => {
+  const { challenge, network, address } = config;
+  console.log(`📡 Downloading ${address} from ${network}...`);
+  const API_URL = `https://api-${network}.etherscan.io/api`;
+  const paramString = `?module=contract&action=getsourcecode&address=${address}&apikey=${ETHERSCAN_API_KEY}`;
+
+  let sourceCodeParsed;
+  try {
+    const response = await axios.get(API_URL + paramString);
+    // The Etherscan API returns OK / NOTOK
+    if (response.data.message !== "OK") {
+      console.log(response.data);
+      throw new Error("Etherscan API returned NOTOK");
+    }
+
+    // On "sourceCode" Etherscan return 3 possible values:
+    // 1. A string (on flattened contracts)
+    // 2. An almost-valid JSON :( (on splitted verified contracts)
+    // 3. A valid JSON (on _some_ splitted verified contracts)
```

Comment:
I saw this in the original code base and I am curious if there is a way to get past it that is cleaner. It can just be a tech debt research item for now but I will be interested to talk with you about it if you have seen it first hand.

Reply:
I copied all the crazy logic for parsing the etherscan source code from `speedrun-grader` because I saw Carlos made the update some 2 months ago. 

I trust Carlos as good dev, so I didnt take the time to fully understand it. Its also hard to battle the edge cases without usage like speedrun ethereum. Maybe good idea to check in with him if you want the full story